src/posix/platform/multicast_routing.cpp

```diff
@@ -0,0 +1,563 @@
+
+    // Enable Multicast Forwarding in Kernel
+    VerifyOrExit(0 == setsockopt(mMulticastRouterSock, IPPROTO_IPV6, MRT6_INIT, &one, sizeof(one)),
+                 error = OT_ERROR_FAILED);
+
+    // Filter all ICMPv6 messages
+    ICMP6_FILTER_SETBLOCKALL(&filter);
+    VerifyOrExit(0 == setsockopt(mMulticastRouterSock, IPPROTO_ICMPV6, ICMP6_FILTER, (void *)&filter, sizeof(filter)),
+                 error = OT_ERROR_FAILED);
+
+    memset(&mif6ctl, 0, sizeof(mif6ctl));
+    mif6ctl.mif6c_flags     = 0;
+    mif6ctl.vifc_threshold  = 1;
+    mif6ctl.vifc_rate_limit = 0;
+
+    // Add Thread network interface to MIF
+    mif6ctl.mif6c_mifi = kMifIndexThread;
+    mif6ctl.mif6c_pifi = if_nametoindex(gNetifName);
+    VerifyOrExit(mif6ctl.mif6c_pifi > 0, error = OT_ERROR_FAILED);
+    VerifyOrExit(0 == setsockopt(mMulticastRouterSock, IPPROTO_IPV6, MRT6_ADD_MIF, &mif6ctl, sizeof(mif6ctl)),
+                 error = OT_ERROR_FAILED);
+
+    // Add Backbone network interface to MIF
+    mif6ctl.mif6c_mifi = kMifIndexBackbone;
+    mif6ctl.mif6c_pifi = if_nametoindex(gBackboneNetifName);
+    VerifyOrExit(mif6ctl.mif6c_pifi > 0, error = OT_ERROR_FAILED);
+    VerifyOrExit(0 == setsockopt(mMulticastRouterSock, IPPROTO_IPV6, MRT6_ADD_MIF, &mif6ctl, sizeof(mif6ctl)),
+                 error = OT_ERROR_FAILED);
+
+exit:
+    if (error != OT_ERROR_NONE)
+    {
+        FinalizeMulticastRouterSock();
+    }
+
+    return error;
+}
+
+void MulticastRoutingManager::FinalizeMulticastRouterSock(void)
+{
+    VerifyOrExit(mMulticastRouterSock != -1);
+
+    close(mMulticastRouterSock);
+    mMulticastRouterSock = -1;
+
+exit:
+    return;
+}
+
+void MulticastRoutingManager::ProcessMulticastRouterMessages(void)
+{
+    otError         error = OT_ERROR_NONE;
+    char            buf[128];
+    int             nr;
+    struct mrt6msg *mrt6msg;
+    Ip6::Address    src, dst;
+
+    nr = read(mMulticastRouterSock, buf, sizeof(buf));
+
+    VerifyOrExit(nr > 0, error = OT_ERROR_FAILED);
+
+    mrt6msg = reinterpret_cast<struct mrt6msg *>(buf);
+
+    VerifyOrExit(mrt6msg->im6_mbz == 0);
+    VerifyOrExit(mrt6msg->im6_msgtype == MRT6MSG_NOCACHE);
+
+    src.SetBytes(mrt6msg->im6_src.s6_addr);
+    dst.SetBytes(mrt6msg->im6_dst.s6_addr);
+
+    error = AddMulticastForwardingCache(src, dst, static_cast<MifIndex>(mrt6msg->im6_mif));
+
+exit:
+    otLogResultPlat(error, "MulticastRoutingManager: %s", __FUNCTION__);
+}
+
+otError MulticastRoutingManager::AddMulticastForwardingCache(const Ip6::Address &aSrcAddr,
+                                                             const Ip6::Address &aGroupAddr,
+                                                             MifIndex            aIif)
+{
+    otError        error = OT_ERROR_NONE;
+    struct mf6cctl mf6cctl;
+    MifIndex       forwardMif = kMifIndexNone;
+
+    VerifyOrExit(aIif == kMifIndexThread || aIif == kMifIndexBackbone, error = OT_ERROR_INVALID_ARGS);
+
+    ExpireMulticastForwardingCache();
```

Comment:
It seems to me that now stale multicast route entries is only removed when there is new multicast traffic?
It might be more reasonable to flush stale entries periodically?

Reply:
`expire per 60s` added